src/main/java/org/cef/CefApp.java

```diff
@@ -23,605 +25,564 @@
+     * are disposed, CefApp will be shutdown.
+     * @param client the disposed client.
+     */
+    final protected void clientWasDisposed(CefClient client) {
+        clients_.remove(client);
+        if (clients_.isEmpty() && getState().compareTo(CefAppState.SHUTTING_DOWN) >= 0) {
+            // Shutdown native system.
+            shutdown();
+        }
     }
-  }
 
-  /**
-   * Shut down the context.
-   */
-  //montoyo: modified for MCEF
-  public final void shutdown() {
-    System.out.println("  shutdown on " + Thread.currentThread());
-
-    // Shutdown message loop
-    try {
-      lock.lock();
-      cefShutdown.signal();
-    } finally {
-      lock.unlock();
+    /**
+     * Initialize the context.
+     * @return true on success.
+     */
+    private final void initialize() {
+        // Execute on the AWT event dispatching thread.
+        try {
+            Runnable r = new Runnable() {
+                @Override
+                public void run() {
+                    String library_path = getJcefLibPath();
+                    System.out.println("initialize on " + Thread.currentThread()
+                            + " with library path " + library_path);
+
+                    CefSettings settings = settings_ != null ? settings_ : new CefSettings();
+
+                    // Avoid to override user values by testing on NULL
+                    if (OS.isMacintosh()) {
+                        if (settings.browser_subprocess_path == null) {
+                            Path path = Paths.get(library_path,
+                                    "../Frameworks/jcef Helper.app/Contents/MacOS/jcef Helper");
+                            settings.browser_subprocess_path =
+                                    path.normalize().toAbsolutePath().toString();
+                        }
+                    } else if (OS.isWindows()) {
+                        if (settings.browser_subprocess_path == null) {
+                            settings.browser_subprocess_path = library_path + "\\jcef_helper.exe";
+                        }
+                    } else if (OS.isLinux()) {
+                        if (settings.browser_subprocess_path == null)
+                            settings.browser_subprocess_path = library_path + "/jcef_helper";
+                        if (settings.resources_dir_path == null)
+                            settings.resources_dir_path = library_path;
+                        if (settings.locales_dir_path == null)
+                            settings.locales_dir_path = library_path + "/locales";
+                    }
+
+                    if (N_Initialize(library_path, appHandler_, settings))
+                        setState(CefAppState.INITIALIZED);
+                }
+            };
+
+            if (SwingUtilities.isEventDispatchThread())
+                r.run();
+            else
+                SwingUtilities.invokeAndWait(r);
```

Comment:
this was another area where I had to call from the other thread or else it would crash in debug, or just fail to run the message loop in the DLL on release.